module/zfs/zpl_super.c

```diff
@@ -333,10 +365,10 @@ const struct super_operations zpl_super_operations = {
 struct file_system_type zpl_fs_type = {
 	.owner			= THIS_MODULE,
 	.name			= ZFS_DRIVER,
-#ifdef HAVE_MOUNT_NODEV
+#ifdef HAVE_FST_MOUNT
 	.mount			= zpl_mount,
 #else
 	.get_sb			= zpl_get_sb,
```

Comment:
It looks like this same issue exists with 2.6.38 and earlier kernels and can basically be fixed in the same way.  How about refactoring the heart of the updated `zpl_mount` in to a new common `zpl_mount_impl()` function which can be used by both `zpl_mount()` and `zpl_get_sb()`.  The only got'cha appears to be that the `zpl_get_sb()` version will need to call `simple_set_mnt()` at the end instead of simply `dget()` and the return types differ.